tf_helpers/tf_dot_general.py

```diff
@@ -0,0 +1,135 @@
+  """ Compose the output string representation.
+
+  e.g., ij, jk, (((1,), (0,)), ((), ())) -> ik
+        aij, ajk, (((2,), (1,)), ((0,), (0,))) -> aik
+
+  Args:
+    lhs_rep: A string representation for the left-hand side input array
+    rhs_rep: A string representation for the right-hand side input array
+    lhs_contraction: Sequence[int] (the contraction dimensions of lhs)
+    rhs_contraction: Sequence[int] (the contraction dimensions of rhs)
+    lhs_batch: Sequence[int] (the batch dimensions of lhs)
+    rhs_batch: Sequence[int] (the batch dimensions of rhs)
+
+  Returns:
+    A string representation of the result array.
+  """
+  output_rep = []
+  for dim in lhs_batch:
+    output_rep.append(lhs_rep[dim])
+
+  for i in _minus(range(len(lhs_rep)), lhs_batch + lhs_contraction):
+    output_rep.append(lhs_rep[i])
+  for i in _minus(range(len(rhs_rep)), rhs_batch + rhs_contraction):
+    output_rep.append(rhs_rep[i])
+  return ''.join(output_rep)
+
+
+def non_batched_matmul(lhs, rhs, lhs_contraction, rhs_contraction):
+  """ Compute the non-batched matrix multiplication.
+
+  If it is the general non-batched/single-batched matrix multiplication,
+  use the highly optimized kernel `tf.tensordot` to handle it.
+
+  Args:
+    lhs: an array (the left-hand side matrix/vector to be multiplied)
+    rhs: an array (the right-hand side matrix/vector to be multiplied)
+    lhs_contraction: Sequence[int] (the contraction dimensions of lhs)
+    rhs_contraction: Sequence[int] (the contraction dimensions of rhs)
+
+  Returns:
+    An array that contains the result.
+  """
+  return tf.tensordot(lhs, rhs, axes=(list(lhs_contraction), list(rhs_contraction)))
+
+
+def tf_dot_general(lhs, rhs, dimension_numbers):
+  """ The general dot operation for TensorFlow.
+
+  An equivalent general dot operation as that in JAX -
+     <https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.dot_general.html>
+  Although there is an implementation in TF XLA, avoid directly using XLA when
+  possible.
+
+  e.g., non-batched: ij,jk->ik
+        batched: ijk,ikl->ijl
+
+  Args:
+    lhs: an array (the left-hand side matrix/vector to be multiplied)
+    rhs: an array (the right-hand side matrix/vector to be multiplied)
+    dimension_numbers: (Tuple[Tuple[Sequence[int], Sequence[int]],
+      Tuple[Sequence[int], Sequence[int]]]) – a tuple of tuples of the form
+      ((lhs_contracting_dims, rhs_contracting_dims), (lhs_batch_dims, rhs_batch_dims))
+
+  Returns:
+    An array that contains the result.
+  """
+  char_list = list(string.ascii_lowercase)
+  char_list = char_list[8:] + char_list[:8]
+  lhs_rank, rhs_rank = len(lhs.shape), len(rhs.shape)
+  lhs_rep = char_list[:lhs_rank]
+  rhs_rep = char_list[lhs_rank:lhs_rank+rhs_rank]
+  contraction, batch = dimension_numbers
+  lhs_contraction, rhs_contraction = contraction
+  if len(lhs_contraction) != len(rhs_contraction):
+    raise ValueError("The input matrices are required to have the same number "
```

Comment:
Minor nit: in NT we usually use single quotes `'` vs `"` for strings, and python 3.6 literal string interpolation for brevity, so here and elsewhere I'd suggest something like
```
raise ValueError('The input matrices are required to have the same number '
                 'of contraction dimensions, but got: '
                 f'lhs {len(lhs_contraction)}, rhs: {len(rhs_contraction)}.')
```

Reply:
Solved in https://github.com/google/neural-tangents/pull/63/commits/bee716ff436936f93473f7df971920e043b53065